embedded-service/src/cfu/component.rs

```diff
@@ -0,0 +1,315 @@
+//! Device struct and methods for component communication
+use core::future::Future;
+use core::ops::DerefMut;
+
+use embassy_sync::blocking_mutex::raw::NoopRawMutex;
+use embassy_sync::channel::Channel;
+use embassy_sync::mutex::Mutex;
+use embedded_cfu_protocol::components::{CfuComponentInfo, CfuComponentStorage, CfuComponentTraits};
+use embedded_cfu_protocol::protocol_definitions::*;
+use embedded_cfu_protocol::{CfuWriter, CfuWriterError, CfuWriterMock};
+use heapless::Vec;
+
+use crate::intrusive_list;
+
+/// Component internal update state
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum ComponentState {
+    /// Component not currently processing an update
+    Idle,
+    /// Component is ready to receive an offer,
+    Ready,
+    /// Component is busy with an update
+    Busy,
+    /// Component has received all new fw bytes and needs finalization logic
+    FinalizingUpdate,
+}
+
+/// Internal device state for power policy
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct InternalState {
+    /// Current state of the device
+    pub state: ComponentState,
+    /// Current consumer capability
+    pub waiting_on_subs: bool,
+}
+
+impl InternalState {
+    /// Constructor for a given `state`
+    pub fn new(state: ComponentState) -> Self {
+        Self {
+            state,
+            waiting_on_subs: false,
+        }
+    }
+    /// Constructor that uses given values for both `state` and `waiting_on_subs`
+    pub fn new_with_subcomponent_info(state: ComponentState, waiting_on_subs: bool) -> Self {
+        Self { state, waiting_on_subs }
+    }
+}
+
+impl Default for InternalState {
+    fn default() -> Self {
+        Self::new(ComponentState::Idle)
+    }
+}
+
+/// CFU Request types and necessary data
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum RequestData {
+    /// Request for component's current FW version
+    FwVersionRequest,
+    /// Contains an offer for the component to evaluate
+    GiveOffer(FwUpdateOfferCommand),
+    /// Contains bytes for an accepted fw offer
+    GiveContent(FwUpdateContentCommand<&'static [u8]>),
```

Comment:
This is where you'll need to start using `SharedRef`. But I'm not sure what the best way to make this generic with other slice is. The protocol crate might have to define its own trait that `SharedRef` implements. I think we could also provide an implementation of that trait for `RefCell` too.

Reply:
for this we know that the data length is going to be up to 52 bytes and any unused bytes are zero-padded so how about I change from using the generic T: AsRef<[u8]> to this:


```suggestion
    pub struct FwUpdateContentCommand> {
        pub header: FwUpdateContentHeader,
        pub data: [u8; DEFAULTDATALENGTH],
    }
```

Reply:
Sounds good to me.